torch_geometric/explain/algorithm/pgm_explainer.py

```diff
@@ -0,0 +1,473 @@
+            p_values.append(p)
+
+        # the original code uses number_candidates_nodes = int(top_nodes * 4)
+        # if we consider 'top nodes' to equate to max number of nodes
+        # it seems more correct to limit number_candidates_nodes to this
+        candidate_nodes = np.argpartition(
+            p_values, max_subgraph_size)[0:max_subgraph_size]
+
+        # Round 2
+        samples = self.batch_perturb_features_on_node(
+            num_samples=num_samples, indices_to_perturb=candidate_nodes, x=x,
+            model=model, **kwargs)
+
+        # note: the PC estimator is in the original code, ie. est= PC(data)
+        # but as it does nothing it is not included here
+        data = pd.DataFrame(np.array(samples.detach().cpu()))
+
+        p_values = []
+        dependent_nodes = []
+
+        target = num_nodes
+        for node in range(num_nodes):
+            chi2, p, _ = chi_square(node, target, [], data, boolean=False,
+                                    significance_level=significance_threshold)
+            p_values.append(p)
+            if p < significance_threshold:
+                dependent_nodes.append(node)
+
+        top_p = np.min((max_subgraph_size, num_nodes - 1))
+        ind_top_p = np.argpartition(p_values, top_p)[0:top_p]
+        pgm_nodes = list(ind_top_p)
+        pgm_stats = {
+            k: v
+            for k, v in dict(zip(range(num_nodes), p_values)).items()
+            if k in candidate_nodes
+        }
+
+        return pgm_nodes, pgm_stats
+
+    def explain_node(
+        self,
+        model: torch.nn.Module,
+        x: Tensor,
+        node_index: int,
+        edge_index: Tensor,
+        edge_weight: Tensor,
+        target: Tensor,
+        num_samples: int = 100,
+        max_subgraph_size: int = None,
+        significance_threshold=0.05,
+        pred_threshold=0.1,
+    ) -> Tuple[List, Dict]:
+        r"""
+        Generate explanations for node classification tasks
+        Args:
+            model (torch.nn.Module): model that generated the predictions
+            x (torch.Tensor): node feature matrix
+            node_index:
+            edge_index (torch.Tensor): edge_index of the input graph
+            edge_weight (torch.Tensor): apply weighting to edges (Optional)
+            target (torch.Tensor):  the prediction labels
+            num_samples (int): num of samples of perturbations used to test
+                the significance of neighbour nodes to the prediction
+            max_subgraph_size:
+            significance_threshold (float): threshold below which to consider
+                a node has an effect on the prediction
+            pred_threshold:
+
+        Returns:
+            pgm_nodes (List): neighbour nodes that are significant
+                 in the selected node's prediction
+            pgm_stats (Dict): node index and their p-values in the
+                model prediction
```

Comment:
Can we combine this into one `Tensor` , which would be a list of `pvalues`. The `pvalues` of nodes outside the neighborhood or insignificant pvalues would be zeroed out.